src/librustc_typeck/check/mod.rs

```diff
@@ -3763,8 +3763,36 @@ pub fn resolve_ty_and_def_ufcs<'a, 'b, 'tcx>(fcx: &FnCtxt<'b, 'tcx>,
                                                         &'a [ast::PathSegment],
                                                         def::Def)>
 {
+
+    // Associated constants can't depend on generic types.
+    fn have_disallowed_generic_consts<'a, 'tcx>(fcx: &FnCtxt<'a, 'tcx>,
+                                                def: def::Def,
+                                                ty: Ty<'tcx>,
+                                                span: Span,
+                                                node_id: ast::NodeId) -> bool {
+        match def {
+            def::DefAssociatedConst(..) => {
+                if ty::type_has_params(ty) || ty::type_has_self(ty) {
```

Comment:
Are you sure that lifetime parameters are relevant? The reason that type parameters are relevant is that they can influence the constant value that you end up with. I don't see how this could be true for lifetime parameters. E.g. I think that if `<&'a Foo>::CONST` can be resolved at all, it should resolve to the same thing as `<&'b Foo>::CONST` and `<&'static Foo>::CONST`?

Hmm, except if you have a trait impl that applies to `<&'a Foo>::CONST`, and an inherent impl that applies only `<&'static Foo>::CONST`, which is *sort of* like a specialization for `'static`? Is that what you mean?